### ISSUE TITLE
Allow arbitrary types for metric states

### ISSUE THREAD

**Issue by LarsHill** [PyTorchLightning/metrics#987]
## 🚀 Feature

Metric states seem to be limited to `torch.Tensor` or `List[torch.Tensor]`.

In my usecase i want to store a dictionary as state. My dataset comprises of samples who can be assigned to different documents. In order to calculate macro metrics (calculate metrics per document and average) I want to store my metric states (e.g. true positives, false positives, etc.) as a dictionary. Here is some pseudocode:

```python
class MyMetric(Metric):

    def __init__(self, dist_sync_on_step: bool = False):
        super().__init__(dist_sync_on_step=dist_sync_on_step)

        self.add_state("statistics", default=defaultdict(lambda: defaultdict(float)), dist_reduce_fx=None)

    def update(
            self,
            predictions: torch.Tensor,
            targets: torch.Tensor,
            document_ids: List
    ):
        predictions = predictions.bool()
        targets = targets.bool()

        tps = predictions * targets
        tns = predictions.logical_not() * targets.logical_not()
        fps = predictions * targets.logical_not()
        fns = predictions.logical_not() * targets

        for id_, tp, tn, fp, fn in zip(document_ids, tps, tns, fps, fns):
            self.statistics[id_]['tp'] += tp.float().item()
            self.statistics[id_]['tn'] += tn.float().item()
            self.statistics[id_]['fp'] += fp.float().item()
            self.statistics[id_]['fn'] += fn.float().item()

    def compute(self):
            ...
```

Unfortunately the above code is not allowed. Each metric state has to be a `torch.Tensor` or a `List[torch.Tensor]`.
That means normal float values or numpy arrays cannot be used as metrics either. Is there a particular reason for that?

**Comment by LarsHill** [comment]
Hey,

Are there any updates on this matter? After checking `metric.py` I don't see a particular reason for this restriction but maybe I overlooked something. 
Also, the `reset` method could simply reset each state variable to the user defined default instead of checking the type to distinguish between `List` and `torch.Tensor`. Regarding reduction logic one can keep the current approach and if the state type is something custom (e.g. a `Dict`, etc.) the synced state is simply a `List[Dict]` and the user is in charge of providing a custome reduce fn to combine the different states across processes.

**Comment by amorehead** [comment]
@LarsHill, I agree that this is an important question to ask. I am also encountering a use case where I need to track my custom `Metrics` state with a dictionary. Any thoughts on this query, @SkafteNicki or @Borda?

**Comment by LarsHill** [comment]
Once again a short ping. Are there any plans to tackle this issue in the near future?

**Comment by SkafteNicki** [comment]
Hi @LarsHill,
It has not really been a priority for us, but I am willing to listen. At the moment I think we could consider adding the option to also support dicts of tensors as an third option. It would not completely solve your initial example, because it requires dict of dict of tensors, but then you could do something like:
```python
for i in range(num_document_ids):
    self.add_state(f"statistics_{i}", default=defaultdict(), dist_reduce_fx=None)
```
and then in update:
```
for id_, tp, tn, fp, fn in zip(document_ids, tps, tns, fps, fns):
    getattr(self, f"statistics_{id_})['tp'] += tp.float().item()
    getattr(self, f"statistics_{id_})['tn'] += tn.float().item()
    getattr(self, f"statistics_{id_})['fp'] += fp.float().item()
    getattr(self, f"statistics_{id_})['fn'] += fn.float().item()
```
this would of cause mean that the `num_documents_ids` needed to be known in advance.

**Comment by LarsHill** [comment]
Hi @SkafteNicki,

Thanks for your suggestion. That would work as a temporary solution, but I think it's not so elegant to iterate the entire dataset at initialization time, just to retrieve all the document ids.

Is there a particular reason, why the state types are restricted in the first place? What about the suggestion to extend it to arbitrary types and if it is something custom, the synced/reduced state is simply a List of that custom type and the user is in charge to properly reduce it, i.e. by providing a specific reduce function. 

For current users that work with `torch.Tensor` oder `List[torch.Tensor]` nothing changes, and users who have to use custom metric state types have the flexibility to do so.

**Comment by 14renus** [comment]
Just wondering,
 is there a reason to not just keep track of a dictionary of metrics as a class variable and override reset() to iterate through the dictionary and call reset?

**Comment by 14renus** [comment]
Actually to answer my own question, I found that internal states are not handled properly by just overridding reset() yourself (which is also stated in the documentation)

forward() (which calls update, reset, update in this order)  doesn't work as expected (the state gets reset() everytime). This is due to the fact that internal metric._copy_state_dict() cannot properly find internal states to copy before calling reset.
